<?xml version='1.0' encoding='utf-8'?>
<robot xmlns:xacro="http://ros.org/wiki/xacro" name="ModularBot">

    <xacro:property name="width" value="0.1" />
    
    <xacro:property name="len1" value="0.4" />
    <xacro:property name="len2" value="0.6" />
    <xacro:property name="len3" value="0.8" />
    <xacro:property name="pi" value="3.1415" />

    <material name="blue">
        <color rgba="0 0 0.8 1" />
    </material>

    <material name="black">
        <color rgba="0 0 0 1" />
    </material>

    <material name="white">
        <color rgba="1 1 1 1" />
    </material>

    <xacro:macro name="default_inertial" params="mass">
        <inertial>
        <mass value="${mass}" />
        <inertia ixx="1.0" ixy="0.0" ixz="0.0" iyy="1.0" iyz="0.0" izz="1.0" />
        </inertial>
    </xacro:macro>

    <link name="base_link" />
    
    <link name="link_0">
        <visual>
            <origin rpy="0 0 0" xyz="0 0 ${len1/2}" />
            <geometry>
                <cylinder length="${len1}" radius="${width}" />
            </geometry>
            <material name="black" />
        </visual>
        <collision>
            <geometry>
                <cylinder length="${len1}" radius="${width}" />
            </geometry>
        </collision>
        <xacro:default_inertial mass="10" />
    </link>

    <joint name="fixed" type="fixed">
        <parent link="base_link" />
        <child link="link_0" />
    </joint>

    <xacro:macro name="add_link" params="suffix length a_i d_i alpha_i theta_i">

        <link name="link_${suffix}">
        <visual>
            <origin rpy="0 0 0" xyz="${length/2} 0 0" />
            <geometry>
            <box size="${length} ${width} ${width}" />
            </geometry>
            <xacro:if value="${suffix%2==0}">
                <material name="white" />
            </xacro:if>
            <xacro:unless value="${suffix%2==0}">
                <material name="blue" />
            </xacro:unless>
        </visual>
        <collision>
            <origin rpy="0 0 0" xyz="${length/2} 0 0" />
            <geometry>
            <box size="${length} ${width} ${width}" />
            </geometry>
        </collision>
        <xacro:default_inertial mass="1" />
        </link>

        <joint name="joint_${suffix-1}_${suffix}" type="continuous">
        <axis rpy="0 0 0" xyz="0 0 1" />
        
        
        
        <parent link="link_${suffix-1}" />

        
        <child link="link_${suffix}" />
        <origin rpy="${alpha_i} 0 0" xyz="${a_i} 0 ${d_i}" />
        </joint>

    </xacro:macro>
  
<xacro:add_link a_i="0" alpha_i="0" d_i="${len1}" length="0.8" suffix="1" theta_i="0" /><xacro:add_link a_i="0.8" alpha_i="${-pi/2}" d_i="0" length="0.8" suffix="2" theta_i="0" /><xacro:add_link a_i="0.8" alpha_i="${pi/2}" d_i="0" length="0.8" suffix="3" theta_i="0" /></robot>